pip/pip-429.md

```diff
@@ -0,0 +1,165 @@
+
+The `compacted_out` field was first introduced in the early stages of development through [#1361](https://github.com/apache/pulsar/pull/1361). However, as part of the overall payload buffer, parsing a `SingleMessageMetadata` currently requires decompressing the compacted entry buffer. This process can be resource-intensive, particularly when handling large topics or encrypted messages, leading to potential performance bottlenecks.
+
+# Motivation
+
+Decompressing the payload buffer solely to check whether individual messages have the `compacted_out` field set is both inefficient and restrictive, as it imposes constraints on the payload buffer format. Furthermore, when using a custom topic compaction service, the entry buffer in the compacted ledger may not include a `SingleMessageMetadata` for every single message, adding further complexity to the process.
+
+This challenge is exacerbated when messages are encrypted, as decompression is not possible without the public key required for decryption. This limitation also impacts the current compaction service, as encrypted messages cannot be compacted. Consequently, operations such as the `GetLastMessageId` RPC will fail, resulting in an error similar to the following:
+
+```
+org.apache.pulsar.client.api.PulsarClientException$BrokerMetadataException: The subscription sub of the topic <topic> gets the last message id was failed
+{"errorMsg":"Failed to read last entry of the compacted Ledger Invalid unknown tag type: 3","reqId":3317275583068061944, "remote":"localhost/127.0.0.1:50818", "local":"/127.0.0.1:50823"}
+```
+
+Instead, the expected behavior is to return the last message ID (e.g., `k1 => null` in the previous example).
+
+Another issue arises from the assumption made by the `GetLastMessageId` RPC that the compacted entry's payload buffer must always contain a `SingleMessageMetadata` list. However, this is not always the case. For instance, a custom topic compaction service might write a payload buffer that omits the `SingleMessageMetadata`. In such cases, the `compactedOut` information could instead be stored in the properties of the `MessageMetadata`, but the `GetLastMessageId` RPC will always fail.
+
+The custom topic compaction service has the flexibility to serialize and deserialize the payload buffer in a different format. However, it still depends on the `GetLastMessageId` RPC in the `hasMessageAvailableAsync` and `getLastMessageIdAsync` methods of the `RawReader` to compute the last message ID. This reliance creates a compatibility issue, as the `GetLastMessageId` RPC will fail when working with a payload buffer in a non-standard format, breaking the functionality of these methods.
```

Comment:
I'm thinking, how about introducing a new method `getLastMessageID()` to the TopicCompactionService for this case?

The MessageMetadata change is still reasonable for pulsar format encrypted data.

Reply:
Accidentally I have this idea last year when we're implementing a dummy topic compaction service: https://github.com/BewareMyPower/pulsar/pull/38

It makes sense to me. Let me update it

Reply:
I've updated the changes on `TopicCompactionService`, PTAL again